### ISSUE TITLE
[feature request] auto choose retention policies based on timestamp when querying

### ISSUE THREAD

**Issue by liyichao** [influxdata/influxdb#2625]
Now, select chooses the default retention policy if not specified. It will be better if select auto choose retention policies for the same series based on timestamp when the select statement does not specify retention policy. This will simplify dashboard tools. This is what graphite already does.

If we have to change retention policy when we want older data, it is tedious because we have to edit the dashboard definition. 

When we select from a series, we do not care what retention policies it has, we just want the datapoints.

**Comment by beckettsean** [comment]
A time range alone is not sufficient to identify which retention policy is desired. There is nothing to prevent two series with identical measurement name and tag sets from existing in separate retention policies with overlapping time ranges. Therefore it is not possible for the system to know which series is intended if the retention policy is not provided. 

A workaround for now is to keep all data for a given dashboard in the same retention policy. It does require maintaining multiple dashboards.

**Comment by daviesalex** [comment]
@beckettsean, this is certainly from our POV a fairly important feature and I dont think your workaround really works. Let me give an example of the problem we have - we capture some data every second. Lets say IO blocks out (which is in telegraf). You need data at 1 second granularity for some types of troubleshooting, but in most cases on a graph it would be crazy to worry about 1 second data.

Lets imagine a common case - a dashboard showing all metrics per server. It might default to show the last hour (3600 data points/server). Per day 86400, per month >2.5 million points. Per hour and perhaps per day will just work but nobody in their right mind would attempt to keep metrics at 1 second granularity over a year and then graph them - while InfluxDB can downsample them its going to have to pull a crazy number of metrics from disk for that query (and, in the real world, it would likely be >1 server on a graph; we also have plans to store data in some cases at a small number of microseconds delta). We also have a basic disk space problem - we are already capturing many hundreds of GB of 1s and 10s metrics per day.

The sane pattern is to keep 1 second for 24 hours, 1 minute for a week, 5 minute for a month and once an hour for a  year (or something similar). This is how just about every other system (graphite, Ganglia, etc.) handle it. This we can sort of do with a Continuous query in InfluxDB, to copy the down sampled data to a new database (although we have to delete the 1 second data manually). The problem is now we have a Grafana problem - we can only query either downsampled, or original data, from a single graph. This means that a user who looks at a 1 hour graph (1 second granularity) then zooms out to see the last month and we have to change the database. Which Grafana does not support.

There are two ways to approach this:
1. Teach Grafana about this concept (preferably somehow auto-learning that downsampled data exists in this other place, although more realistically defining that in Grafana)
2. Provide a single "view" inside InfluxDB that merges the various levels of data
3. Provide a way to down-sample data in InfluxDB after a certain period of time, sort of like a contiguous down-sampling job.

My personal preference would be (3), but I suspect thats not an architectural starter (although if you would be willing to accept that as an option, we might be able to find somebody to work on it and send you a PR). This leaves us with (1) or (2). THis ticket strikes me as asking for (1). DO you think its best to attack this via means of this, or to track a issue more like (2) (for the InfluxDB project)

cc @sebito91, @wrigtim

**Comment by PaulKuiper** [comment]
A different way of solving is is to build a proxy between grafana and influx (we need this anyway to check user acces). Parse out the group by, measurement name and agregate of the incomming request at the proxy and apply a rule to change the measurement name (prepend a retention or a custom string fitting your data structure). Send this query to influx instead of the original. I think this is the most pratical solution at the moment.

**Comment by daviesalex** [comment]
@PaulKuiper, funnily enough thats exactly how we plan to achieve this (we also have the ACL problem).

Have you already worked on this? We may build this and open source it... or use somebody elses's if its already out there.

**Comment by PaulKuiper** [comment]
"""
Little Grafana Influx proxy
Author: Paul Kuiper
Free use
Requires: gevent, bottle, requests
"""

import gevent # micro threading framework
from gevent import monkey
monkey.patch_all()

import sys
import requests # http for humans
from bottle import get, abort, run, request, response, redirect  # micro web framework for humans (use flask if you like)
import regex as re # regular expressions libary

signal_set = set()
pattern = re.compile(r"""
    ^                   # beginning of string
    select\b            # must start with select statement (followed by word boundary)
    \s+                 # 1 or more whitespaces
    (count|min|max|mean|sum|first|last) # an aggragate function                         group 0 (aggregate)
    \(                  # time with opening bracket
    (.*)                # the field name                                                group 1 (field name)
    \)                  # closing bracket
    \s+                 # 1 or more whitespaces

    \bfrom\b            # the from statement should follow (with word boundaries)
    \s+                 # 1 or more whitespaces
    (.*)                # the from content                                              group 2  (measurement)
    \s+                 # 1 or more whitespaces

    \bwhere\b           # the where statement is always present in a grafana query
    (.*)                # the where content                                             group 3  (where clause)

    \bgroup\sby\b       # match group by statement
    \s+                 # 1 or more whitespaces
    time\(              # time with opening bracket
    (\d+)               # minimal 1 digit (does not match 0.1s!)                        group 4  (number of time units)
    ([s|m|h|d|w])       # the group by unit                                             group 5  (time unit)
    \)                  # closing bracket
    \s+                 # 1 or more whitespaces

    \border\sasc        # ordering is always present in a grafana query
    $                   # end of string
    """,  re.VERBOSE | re.I)

@get('/<path:path>') # pass through all other queries
def proxy_influx_query(path):
    """
    Capture the query events comming from Grafana.
    Investegate the query and replace the measurement name with a continuous query measurement name if possible.
    Send out the (modified or unmodified) query to Influx and return the result
    """

    forward_url = r'http://localhost:8086'  # The local influx host

    params = dict(request.query) # get all query parameters
    try:
        params['q'] = modify_query(params['q'], signal_set)
    except Exception as e:
        pass

    headers = request.headers
    cookies = request.cookies
    r = requests.get(url=forward_url + path, params=params, headers=headers, cookies=cookies, stream=True) # get data from influx
    if r.status_code == 200:
        for key, value in dict(r.headers).iteritems():
             response.set_header(key, value)
        for key, value in dict(r.cookies).iteritems():
            response.cookies[key] = value
        pass
    else:
        abort(r.status_code, r.reason) # NOK, return error
    return r.raw


def modify_query(qry, signal_set):
    """
    Grafana will zoom out with the following group by times:

    0.1s, 1s, 5s, 10s, 30s, 1m, 5m, 10m, 30m, 1h, 3h, 12h, 1d, 7d, 30d, 1y

    :param qry: the query to analyse and modify if needed
    :return: the modified query
    """

    try:
        items = pattern.search(qry).groups() # get the content of the different query parts
        table = items[2].strip("'").strip('"')
        measurement = '.'.join((table,'1'+items[5], items[0])) # construct agg. measurement name e.g.: metric.1h.max
        if measurement not in signal_set: # not a known signal (continuous query)
            return qry
        field_name = items[0] + '(' + items[1] + ')'
        group_by = 'time(' + str(items[4]) + items[5] + ')'
        qry = ' '.join(('select',field_name,'from','"'+ measurement + '"','where',items[3],'group by',group_by,'order asc'))
    except Exception as e:
        print e
        pass
    return qry

def get_signals(signal_set):
    params = {  'q' : 'list series',        # get all signals
                'p' : 'GrafanaPassword1!',
                'u' : 'Grafana'}
    while True:
        r = requests.get(r'http://euv-dashboard.eu.asml.com/data//db/db/series', params=params)
        r = r.json()[0]['points']           # convert the json to an object and get the data points
        signal_set =  {sig[1] for sig in r}  # create a set with the signal names only
        gevent.sleep(3600)                  # refresh signal list every hour

if __name__ == '__main__':
    print >> sys.stderr, "Starting signal list updates"
    gevent.spawn(get_signals, signal_set) # start the signal list refresh service
    print >> sys.stderr, "Starting proxy server"
    run(host='0.0.0.0', port=3004, server='gevent') # start the proxy webserver (choose your own port)

**Comment by PaulKuiper** [comment]
Attached is a python file (in txt format, else I could not upload it), which you can use as a simple proxy between grafana and influx.

It can greatly increase zoom speed. It assumes that the following continous queries are present for the measurement called "metric" :

metric.1s.max
metric.1m.max
metric.1h.max
metric.1d.max
metric.1h.mean 
......

Point your "data source" to port 3004 (or whatever you choose) instead of port 8086 in Grafana. 
The proxy will now change your query transparantly by choosing a different table when zooming out.
select max(value) from "metric" where time > x1 and time < x2 group by time(12h)
becomes:
select max(value) from "metric.1h.max" where time > x1 and time < x2 group by time(12h)

[poxy.txt](https://github.com/influxdb/influxdb/files/51072/poxy.txt)

**Comment by huhongbo** [comment]
+1

**Comment by toni-moreno** [comment]
+1

**Comment by exeral** [comment]
+1 !!
any feedback about PaulKuiper workaround ?

**Comment by PaulKuiper** [comment]
I'll update it for influxdb 0.10 somewhere this month

**Comment by adrianlzt** [comment]
+1

**Comment by thbourlove** [comment]
:+1:

**Comment by Lupul** [comment]
I've done some work on @PaulKuiper proxy to work with 0.10 and put it here:
https://github.com/Lupul/influxdb-grafana-rp-proxy

**Comment by adrianlzt** [comment]
With version 0.10 it's normal to have several values in each measurement.
In the proxy readme the CQs are just for only one value (called _value_)
```
CREATE CONTINUOUS QUERY graphite_cq_10sec  ON graphite BEGIN SELECT mean(value) as value INTO graphite."10sec".:MEASUREMENT  FROM graphite."default"./.*/ GROUP BY time(10s), * END
```
Any ideas how to handle where there are several values?

I was thinking in some batch processing with kapacitor which obtains all values for each measurement, and creates the appropriates CQs.

**Comment by adrianlzt** [comment]
I have make a small script to autogenerate RPs and CQs: https://gist.github.com/f4b6f5c8f6c2a51c3f60

**Comment by beckettsean** [comment]
@adrianlzt in CQs each tag or field must be explicitly named. It is possible to use `SELECT *` to return all columns from an ad hoc query, but not in a CQ, as there is no aggregation function.

So, to downsample multiple fields, the CQ would look something like this:
```
CREATE CONTINUOUS QUERY graphite_cq_10sec  
ON graphite BEGIN 
SELECT mean(value) as value, last(value) as last, mean(value_23) as value_23, top(field19) as top
INTO graphite."10sec".:MEASUREMENT  
FROM graphite."default"./.*/ 
GROUP BY time(10s), * 
END
```

**Comment by adrianlzt** [comment]
Will be this fixed in next versions?
It is hard to maintain downsampling in multiple databases with lots of
series with multiple values.

**Comment by beckettsean** [comment]
Follow https://github.com/influxdata/influxdb/issues/5750, which is the
relevant issue.

On Fri, Mar 4, 2016 at 9:13 AM, Adrián López <notifications@github.com>
wrote:

> Will be this fixed in next versions?
> It is hard to maintain downsampling in multiple databases with lots of
> series with multiple values.
>
> El vie., 4 de marzo de 2016 2:20, Sean Beckett <notifications@github.com>
> escribió:
>
> > @adrianlzt <https://github.com/adrianlzt> in CQs each tag or field must
> > be explicitly named. It is possible to use SELECT * to return all columns
> > from an ad hoc query, but not in a CQ, as there is no aggregation
> function.
> >
> > So, to downsample multiple fields, the CQ would look something like this:
> >
> > CREATE CONTINUOUS QUERY graphite_cq_10sec
> > ON graphite BEGIN
> > SELECT mean(value) as value, last(value) as last, mean(value_23) as
> value_23, top(field19) as top
> > INTO graphite."10sec".:MEASUREMENT
> > FROM graphite."default"./.*/
> > GROUP BY time(10s), *
> > END
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/influxdata/influxdb/issues/2625#issuecomment-192046509>
> > .
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/influxdata/influxdb/issues/2625#issuecomment-192362529>
> .
>



-- 
Sean Beckett
Director of Support and Professional Services
InfluxDB

**Comment by nelg** [comment]
+1 for this issue.  Without good data down-sampling / roll up, moving from whisper is likely to present problems for graphite graphs over long time periods.  Ideally needs a way to set a default policy around rollup and retention, for all metrics of a specific type.

**Comment by TomGudman** [comment]
+1

Coming from the old RRD world, this is obviously a big shift in mentality, I fully agree with https://github.com/influxdata/influxdb/issues/2625#issuecomment-159263547 - option 3

Then I never used graphite but read about it several times and I liked how you can set different retention policies per metrics if desired otherwise you get the default downsampling automatically. 

The influxdb approach is awkward and makes it hard to maintain in my beginner's opinion.

I believe users want a time series databases that is efficient, fast and requires low maintenance. Managing RP and CQ with complex InfluxQL queries isn't obvious... 

My comments maybe irrelevant, I am still learning and reading mailing-list and github issues to figure out how I can set downsampling. Currently, I start to wonder why I waited so long for InfluxDB instead of just using Dixon's Graphite. 

Still, influxdb is fast to spin and play with the tutorial but then it gets more complicated when you really want to do something with it.

**Comment by jsternberg** [comment]
Looking through old issues and I found this one. It seems related to #6910.

**Comment by daviesalex** [comment]
Agreed. I think this one can basically be closed as a dup of #6910

**Comment by jsternberg** [comment]
I'm going to close this in favor of #6910. If a new issue gets created for this, it will be mentioned in that issue.
